Anonymization decrypt

constcrist opened the pull request:
no unit tests. some unit tests are needed to test at least JSON decoding...

constcrist committed:
adds dependency modules

constcrist committed:
supports JSON response version 2

- see https://github.com/intuitivelabs/intuitive-issues/issues/100 for
  details on JSON response format

constcrist committed:
adds support for encrypted IPs

jiriatipteldotorg commented on clients/go/apiban-ipsets/apiban-ipsets.go:
```diff
@@ -203,6 +209,16 @@ func main() {
+			log.Fatalln("Cannot initialize ipcipher. Exiting.")
+		}
+		// initialize a validator using the configured passphrase; neither length nor salt are used since this validator verifies only the remote code
+		if validator, err = anonymization.NewPassphraseValidator(apiconfig.PASSPHRASE, 0 /*length*/, "" /*salt*/); err != nil {
```
probably a silly question but how can we validate the checksum when the validation gets only the password as input without the UID? Do we take the UID from the event's UDI:HMAC pair? Does it mean we calculate it for every event?  If so, wouldn't calculating it once and doing `memcmp` only to validate be sufficient?

jiriatipteldotorg commented on clients/go/apiban-ipsets/apiban-ipsets.go:
```diff
@@ -293,11 +310,38 @@ func run(ctx context.Context, blset ipset.IPSet, apiconfig ApibanConfig) error {
+						if !encryptionOn {
+							log.Print("IP encrypted but no passphrase configured")
+							continue
+						}
```
Also as I'm not literate in `go` -- what does exactly happen when password is configured but a password-less event is encountered -- it seems the IP address is skipped? I would vote for being liberal, issuing a warning and including it in IPSET. Motivation: we could then more easily mix up user's IP addresses with some public shared IP address list, like honeynet.

constcrist committed:
adds support for version 2 JSON

constcrist committed:
moves response processing in apiban.go

constcrist commented:
JSON ver 2 processing is supported now. I have updated all the unit tests to reflect this change and I tested the decoder with JSON objects provided here:
https://github.com/intuitivelabs/intuitive-issues/issues/100
Internal state processing is still a mess (ID is stored in some structure loaded from configuration and updated back in the configuration on the disk)
the overall software architecture is a mess. re-writing from scratch may take some time though...

constcrist committed:
adds profiling to a DUMMY version

- most of the changes here have to be discarded

constcrist committed:
splits into multiple src files, adds fs state

constcrist committed:
turns profiling off

constcrist committed:
adds global var to ctrl load/save state

constcrist committed:
adds password, changes url

constcrist committed:
adds encryption key

constcrist committed:
removes query loop

- only one GET request is sent per select timeout

constcrist committed:
implements encryption in separate functions

constcrist committed:
adds new config.json file for siem server

constcrist committed:
updates anonymization dependency
